test/testdata/infer/metatype_new.rb

```diff
@@ -0,0 +1,37 @@
+# typed: true
+
+class MyGeneric
+  extend T::Generic
+
+  X = type_member
+end
+
+MyArrayString = T.type_alias {T::Array[String]}
+
+# -- good --
+
+String.new
+T::Array[String].new
+MyGeneric.new
+MyGeneric[Integer].new
+
+# -- bad, but we can't detect them yet --
+
+MyArrayString.new                  # Type alias is unwound by the type we get to MetaType::dispatchCall
+
+# -- bad --
+
+T.any(Integer, String).new         # error: Cannot call `new` on type `T.any(Integer, String)`
+T.all(Kernel, Comparable).new      # error: Cannot call `new` on type `T.all(Kernel, Comparable)`
+T.nilable(String).new              # error: Cannot call `new` on type `T.nilable(String)`
+
+T.untyped.new                      # error: Cannot call `new` on type `T.untyped`
+T.self_type.new                    # error: Cannot call `new` on type `T.untyped`
+
+T.class_of(String).new             # error: Cannot call `new` on type `T.class_of(String)`
+T.class_of(MyGeneric).new          # error: Cannot call `new` on type `T.class_of(MyGeneric)`
+T.class_of(MyGeneric[Integer]).new # error: Cannot call `new` on type `T.class_of(MyGeneric)`
+
+T.proc.void.new                    # error: Cannot call `new` on declaration builder type
```

Comment:
I didn't see code to handle this case. How does this work?

Reply:
This gets handled in `dispatchCallSymbol`. It doesn't seem to go through `MetaType::dispatchCall`.

Reply:
Sounds good, thanks!

You've found some more bugs:

```ruby
# typed: true
extend T::Sig

fs = T::Array[T.proc.params(x: Integer).void].new
T.reveal_type(fs)

fs.each do |f|
  f.call(0)
  f.call('')
end

# ---

g = T.let(->(x) {}, T.proc.params(x: Integer).void)
T.reveal_type(g)

g.call(0)
g.call('')
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Afs%20%3D%20T%3A%3AArray%5BT.proc.params(x%3A%20Integer).void%5D.new%0AT.reveal_type(fs)%0A%0Afs.each%20do%20%7Cf%7C%0A%20%20f.call(0)%0A%20%20f.call('')%0Aend%0A%0A%23%20---%0A%0Ag%20%3D%20T.let(-%3E(x)%20%7B%7D%2C%20T.proc.params(x%3A%20Integer).void)%0AT.reveal_type(g)%0A%0Ag.call(0)%0Ag.call(''))

The code above and below the line should behave the same.

Reply:
You mean `g` (where the `call` is allowed if the arg types match) is the correct behavior and `f` is the one that's bugged, right?

Reply:
Yes, the `g` code is correct, and the `f` code should behave like the `g` code.

Reply:
Does #4378 capture it correctly?

Reply:
yep!